source/_posts/2016-01-05-improvements-to-error-responses-in-rails-5-api-mode.markdown

```diff
@@ -0,0 +1,160 @@
+    posts GET    /posts(.:format)          posts#index
+          POST   /posts(.:format)          posts#create
+ new_post GET    /posts/new(.:format)      posts#new
+edit_post GET    /posts/:id/edit(.:format) posts#edit
+     post GET    /posts/:id(.:format)      posts#show
+          PATCH  /posts/:id(.:format)      posts#update
+          PUT    /posts/:id(.:format)      posts#update
+          DELETE /posts/:id(.:format)      posts#destroy
+```
+
+These routes are what you get when you have something like `resources: :post` in your routes file. We can see that _format_ here is a dynamic segment present in every generated routes, and this is the most common way in Rails to decide which is the format used to form a response. So, in many cases the format depends on the extension in the request URL, usually appended by a client application (or client side code if it's a Rails default stack application) to indicate the expected format.
+
+Therefore, a JSON response is generated if you issue a GET request to `http:://mysite.com/posts.json` and HTML response is generated when the URL is `http:://mysite.com/posts.html`. In the case of the `index` action, it just relies on the ability of Rails to render the list of posts in the format value inferred from the route path. If the format is not specified in the request path or it's an unknown format for Rails, the response will be rendered in HTML format if it's a standard Rails application or JSON if it's configured with `config.api_only = true` (using the new API mode).
+
+By using `respond_to`, like in the 'create' action of our `PostController`, you can be specific about the content of the response depending on the specified format. In this case, if the format is not provided it will use the first format listed within the `respond_to` block. If there is not a handler for the specified format, it will raise a `ActionController::UnknownFormat` exception (and it would result in a `406 Not Acceptable` response error).
+
+In general terms, this is a game of two players: we have a client that makes a request to our server and it could *specify* the expected response format and we also have our application server that is *able to respond* in one or more formats. Therefore, the response format should be determined based on what was requested but also considering the list of supported formats in the server. In Rails, adding the format in the URL is the most common way to specify the expected format in the response. This mechanism relies on how things are defined in the routes file and the controller code, as it was described above.
+
+## Defining the format with the HTTP Accept header
+
+In the context of an API application, I personally prefer the use of the HTTP `Accept` header which is also supported in Rails, because this is the manner to specify the response format according to the HTTP protocol. The [HTTP/1.1 RFC](http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html) states:
```

Comment:
Now you mention that this is supported but reading it top-down when you explain how the format is determined in the previous section it states that only uses the path.